File: tests/clean_install_smoke.py
"""Cross-platform smoke test against the built wheel, never editable source."""

from __future__ import annotations

import shutil
import subprocess
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
wheel = next((ROOT / "dist").glob("*.whl"), None)
if wheel is None:
    raise SystemExit("built wheel not found")

subprocess.run(
    [sys.executable, "-m", "pip", "install", "--force-reinstall", str(wheel)],
    check=True,
)

command = shutil.which("gc-validate")
if command is None:
    raise SystemExit("installed gc-validate console script not found on PATH")


def expect(code: int, *args: str) -> None:
    result = subprocess.run([command, *args], text=True, capture_output=True)
    if result.returncode != code:
        raise SystemExit(
            f"expected exit {code}, got {result.returncode}\n"
            f"stdout:\n{result.stdout}\nstderr:\n{result.stderr}"
        )


fixtures = ROOT / "tests" / "fixtures"
expect(0, "--spec", "ons", "--output", "json", str(fixtures / "valid checks ünicode.json"))
expect(1, "--spec", "ons", str(fixtures / "invalid checks ünicode.json"))
expect(2, "--spec", "ons", str(fixtures / "missing file.json"))

File: tests/clean-install-smoke.mjs
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const fixtures = join(dirname(fileURLToPath(import.meta.url)), "fixtures");
const command = process.platform === "win32" ? "gc-validate.cmd" : "gc-validate";

function windowsQuote(value) {
  return `"${value.replaceAll('"', '""')}"`;
}

function expect(code, args) {
  const result = process.platform === "win32"
    ? spawnSync(process.env.ComSpec || "cmd.exe", ["/d", "/s", "/c", [command, ...args.map(windowsQuote)].join(" ")], { encoding: "utf8" })
    : spawnSync(command, args, { encoding: "utf8", shell: false });
  if (result.error) throw result.error;
  if (result.status !== code) {
    throw new Error(`expected exit ${code}, got ${result.status}\nstdout:\n${result.stdout}\nstderr:\n${result.stderr}`);
  }
}

expect(0, ["--spec", "ons", "--output", "json", join(fixtures, "valid checks ünicode.json")]);
expect(1, ["--spec", "ons", join(fixtures, "invalid checks ünicode.json")]);
expect(2, ["--spec", "ons", join(fixtures, "missing file.json")]);
